<a href="https://colab.research.google.com/github/Only-Eni/Google-Developers-Group-Campus/blob/main/GDG_AI_DAY5_PANDAS_NUMPY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv("data.csv")

df.head()


Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
3        45  '2020/12/04'    109       175     282.4
4        45  '2020/12/05'    117       148     406.0

# 🔍 Key Insights from the Dataset

### 1. Routine Consistency

Approximately 74% of the workouts lasted exactly 60 minutes.

This indicates a highly disciplined and structured exercise habit. Rather than random or sporadic activity, the data reflects a deliberate commitment to standardized training sessions.

Consistency is often a key predictor of long-term fitness progress.

In [ ]:
df.groupby("Duration")["Calories"].mean()

Duration
30     195.10000
45     291.88000
60     314.46087
450    253.30000
Name: Calories, dtype: float64

### 2. Duration Drives Calories

A clear relationship exists between workout duration and calories burned.

- 30-minute sessions average ~195 kcal
- 60-minute sessions average ~317 kcal

This reinforces a fundamental physiological principle: energy expenditure increases with time under activity.

However, duration alone does not fully explain calorie variation — which leads to the next insight.

In [ ]:
df["Pulse"].mean()

np.float64(103.5)

### 3. Aerobic Zone Training

The overall average pulse across sessions is approximately 103 bpm.

This falls within the aerobic training zone for many adults, suggesting that the workouts were designed for cardiovascular endurance rather than maximum intensity or anaerobic bursts.

This reflects a focus on sustainable cardio conditioning rather than high-intensity intervals.

In [ ]:
df[df["Pulse"] > df["Maxpulse"]]

Duration          Date  Pulse  Maxpulse  Calories
23        60  '2020/12/23'    130       101     300.0

### 4. Data Quality Anomaly

One session (December 23) contains a physiologically impossible entry:
the recorded average pulse (130 bpm) exceeds the maximum pulse (101 bpm), This inconsistency highlights the importance of data validation before analysis. Even fitness tracking devices or exports can contain recording or entry errors.

Without identifying such anomalies, conclusions drawn from the dataset could be misleading.

In [ ]:
df[df["Duration"] == 60]["Calories"].min()

df[df["Duration"] == 60]["Calories"].max()

479.0

In [ ]:
df.groupby("Pulse")["Calories"].mean()
df[["Pulse","Calories"]].corr().corr()

Pulse  Calories
Pulse       1.0      -1.0
Calories   -1.0       1.0

### 5. pulse Intensity Influences Calorie Burn

Even among 60-minute sessions, calories burned range from 215 kcal to 479 kcal — more than double.

This indicates that intensity, reflected by heart rate variation, significantly impacts energy expenditure.

Among sessions of equal duration, calorie burn increases with higher recorded pulse, suggesting a positive association between workout intensity and energy expenditure

## Overall Explanation

This dataset tells the story of a consistent fitness routine centered around  60-minute  sessions.

While the time invested is a major contributor to calorie expenditure, pulse intensity plays a critical role in differentiating workout outcomes.

The presence of a data anomaly also reinforces the importance of validating data before interpretation

Overall, the data reflects structured training behavior, moderate cardiovascular intensity, and meaningful variation driven by exertion levels.